numcodecs/bitround.py

```diff
@@ -0,0 +1,82 @@
+import numpy as np
+
+
+from .abc import Codec
+from .compat import ensure_ndarray, ndarray_copy
+
+max_bits = {
+    "float16": 10,
+    "float32": 23,
+    "float64": 52,
+}
+types = {
+    "float16": np.int16,
+    "float32": np.int32,
+    "float64": np.int64,
+}
+inverse = {
+    "int16": np.float16,
+    "int32": np.float32,
+    "int64": np.float64
+}
+
+
+class BitRound(Codec):
+    """Floating-point bit rounding codec
+
+    Drops a specified number of bits from the floating point mantissa,
+    leaving an array more amenable to compression. The number of bits to keep should
+    be determined by an information analysis of the data to be compressed. See
+    https://github.com/zarr-developers/numcodecs/issues/298 for discussion
+    and the original implementation in Julia referred to at
+    https://www.nature.com/articles/s43588-021-00156-2
+
+    Parameters
+    ----------
+
+    keepbits: int
+        The number of bits of the mantissa to keep. The range allowed
+        depends on the dtype input data. If keepbits is
+        equal to the maximum allowed for the data type, this is equivalent
+        to no transform.
+    """
+
+    codec_id = 'bitround'
+
+    def __init__(self, keepbits: int):
+        if keepbits < 0:
+            raise ValueError("keepbits must be zero or positive")
+        self.keepbits = keepbits
+
+    def encode(self, buf):
+        """Create int array by rounding floating-point data
+
+        The itemsize will be preserved, but the output should be much more
+        compressible.
+        """
+        a = ensure_ndarray(buf)
```

Comment:
@martindurant can this also be implemented somehow differently? This line fails when using `xr.apply_ufunc(dask="allowed")` or `xr.map_blocks`, but it works `xr.apply_ufunc(dask="parallelized")` and eager data. Not critical but would be nice for scaling this function out
```
# bitinformation_pipeline.bitround._xr_bitround is round(data, keepbits) from test_bitround.py
xr.apply_ufunc(bitinformation_pipeline.bitround._xr_bitround, ds,7, dask='allowed')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 xr.apply_ufunc(bitinformation_pipeline.bitround._xr_bitround, ds,7, dask='allowed')

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/xarray/core/computation.py:1159, in apply_ufunc(func, input_core_dims, output_core_dims, exclude_dims, vectorize, join, dataset_join, dataset_fill_value, keep_attrs, kwargs, dask, output_dtypes, output_sizes, meta, dask_gufunc_kwargs, *args)
   1157 # feed DataArray apply_variable_ufunc through apply_dataarray_vfunc
   1158 elif any(isinstance(a, DataArray) for a in args):
-> 1159     return apply_dataarray_vfunc(
   1160         variables_vfunc,
   1161         *args,
   1162         signature=signature,
   1163         join=join,
   1164         exclude_dims=exclude_dims,
   1165         keep_attrs=keep_attrs,
   1166     )
   1167 # feed Variables directly through apply_variable_ufunc
   1168 elif any(isinstance(a, Variable) for a in args):

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/xarray/core/computation.py:286, in apply_dataarray_vfunc(func, signature, join, exclude_dims, keep_attrs, *args)
    281 result_coords = build_output_coords(
    282     args, signature, exclude_dims, combine_attrs=keep_attrs
    283 )
    285 data_vars = [getattr(a, "variable", a) for a in args]
--> 286 result_var = func(*data_vars)
    288 if signature.num_outputs > 1:
    289     out = tuple(
    290         DataArray(variable, coords, name=name, fastpath=True)
    291         for variable, coords in zip(result_var, result_coords)
    292     )

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/xarray/core/computation.py:727, in apply_variable_ufunc(func, signature, exclude_dims, dask, output_dtypes, vectorize, keep_attrs, dask_gufunc_kwargs, *args)
    722     if vectorize:
    723         func = _vectorize(
    724             func, signature, output_dtypes=output_dtypes, exclude_dims=exclude_dims
    725         )
--> 727 result_data = func(*input_data)
    729 if signature.num_outputs == 1:
    730     result_data = (result_data,)

File ~/bitinformation_pipeline/bitinformation_pipeline/bitround.py:12, in _xr_bitround(data, keepbits)
     10 codec = BitRound(keepbits=keepbits)
     11 data = data.copy()  # otherwise overwrites the input
---> 12 encoded = codec.encode(data)
     13 return codec.decode(encoded)

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/numcodecs/bitround.py:57, in BitRound.encode(self, buf)
     51 def encode(self, buf):
     52     """Create int array by rounding floating-point data
     53 
     54     The itemsize will be preserved, but the output should be much more
     55     compressible.
     56     """
---> 57     a = ensure_ndarray(buf)
     58     bits = max_bits[str(a.dtype)]
     59     all_set = np.frombuffer(b"\xff" * a.dtype.itemsize, dtype=types[str(a.dtype)])

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/numcodecs/compat.py:45, in ensure_ndarray(buf)
     39     raise TypeError('array.array with char or unicode type is not supported')
     41 else:
     42 
     43     # N.B., first take a memoryview to make sure that we subsequently create a
     44     # numpy array from a memory buffer with no copy
---> 45     mem = memoryview(buf)
     47     # instantiate array from memoryview, ensures no copy
     48     arr = np.array(mem, copy=False)

TypeError: memoryview: a bytes-like object is required, not 'Array'
```

Reply:
My usecase for this codec is to bitround input netcdf to output netcdf with xr.to_netcdf to disk, so this is a usecase outside of zarr

Reply:
I do not understand the differences in all details, https://stackoverflow.com/questions/51736172/whats-the-difference-between-dask-parallelized-and-dask-allowed-in-xarrays-app

Reply:
Sorry, I'm heading out the door. But at a glance, is this a #305 issue (if so, cc: @madsbk)

Reply:
@aaronspring , updated: does this work for you?

Reply:
The change to `ensure_ndarray_like` still leads to the same problem unfortunately.
```
File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/numcodecs/bitround.py:57, in BitRound.encode(self, buf)
     51 def encode(self, buf):
     52     """Create int array by rounding floating-point data
     53 
     54     The itemsize will be preserved, but the output should be much more
     55     compressible.
     56     """
---> 57     a = ensure_ndarray_like(buf)
     58     if not a.dtype.kind == "f" or a.dtype.itemsize > 8:
     59         raise TypeError("Only float arrays (16-64bit) can be bit-rounded")

File /work/mh0727/m300524/conda-envs/bitinfo/lib/python3.10/site-packages/numcodecs/compat.py:42, in ensure_ndarray_like(buf)
     38     raise TypeError("array.array with char or unicode type is not supported")
     39 else:
     40     # N.B., first take a memoryview to make sure that we subsequently create a
     41     # numpy array from a memory buffer with no copy
---> 42     mem = memoryview(buf)
     43     # instantiate array from memoryview, ensures no copy
     44     buf = np.array(mem, copy=False)

TypeError: memoryview: a bytes-like object is required, not 'DataArray'
```

Reply:
Hm, so not that. Can you try with https://github.com/zarr-developers/numcodecs/pull/299/commits/9622fe318777d769d170f650eaa304dd5c6e5b3b (i.e., don't call `encode()`, call the `_bitround()` function)

Reply:
Oh, but it works with `dask.array.map_blocks` when `import dask.array as da; da.map_blocks(bitinformation_pipeline.bitround._np_bitround, ds[v].data, keepbits=7, dtype="float32").compute()`, likely I use `xr.map_blocks` in a wrong way

Reply:
happens in both `encode()` and `decode()`

Reply:
but as `dask.array.map_blocks` works, I will open an issue in `xarray`. And I am very happy that `xr.apply_ufunc(dask="parallelized")` works out-of-the-box!

Reply:
What is `bitinformation_pipeline`?

Reply:
we are wrapping @milankl's `bitinformation.jl` to be called in `python` and directly apply these functions to `xarray` objects.

Reply:
> Oh, but it works with `dask.array.map_blocks` when `import dask.array as da; da.map_blocks(bitinformation_pipeline.bitround._np_bitround, ds[v].data, keepbits=7, dtype="float32").compute()`, likely I use `xr.map_blocks` in a wrong way

started https://github.com/pydata/xarray/discussions/6482

Reply:
Did you try my `_bitround()` version of this?

Reply:
Yes, it uses this PR. https://github.com/pydata/xarray/discussions/6482 and the docs explain that xr.apply_ufunc(dask="allowed") doesnt work for numpy functions.

Reply:
sorry, my fault dragging you into this...

Reply:
OK, let's mark this resolved then - I'll say that the current state of the code is good.